### PR TITLE
Add a picture gallery to pics.html page, with a carousel to switch between images

### PR DESCRIPTION
**Pull request by PrabSG**
Add working header links to both home and Photos page. Fix previous security issue of using innerHTML and fix the use of var to now use const in script.js.

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -31,16 +31,77 @@ function addRandomGreeting() {
...
+
+  let newIndex = n;
+  if (n >= slides.length) {
+    newIndex = 0;
+  }
+  if (n < 0) {
+    newIndex = slides.length - 1;
+  }
+  
+  for (let i = 0; i < slides.length; i++) {
```
Comment (DarthGandalf) [inline portfolio/src/main/webapp/script.js]: `for (let slide of slides)

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -31,16 +31,77 @@ function addRandomGreeting() {
...
+ */
+function advanceSlides(n) {
+  currIndex += n;
+  currIndex = showSlidePicture(currIndex);
+};
+
+/**
+ * Show the image from the gallery at the given position.
+ */
+function showSlidePicture(n) {
```
Comment (DarthGandalf) [inline portfolio/src/main/webapp/script.js:88]: This function is probably complicated enough to have a unit test

**Review by DarthGandalf — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 1 above)
Comment (PrabSG) [reply portfolio/src/main/webapp/script.js]: Done

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -31,16 +31,77 @@ function addRandomGreeting() {
...
+  const slides = document.getElementsByClassName("slide-pic");
+  const caption = document.getElementById("slide-caption");
+
+  // If not on pics.html document wil not contain any slide elements.
+  if (slides.length === 0) {
+    return n;
+  }
+
+  let newIndex = n;
+  if (n >= slides.length) {
```
Comment (med0b1011) [inline portfolio/src/main/webapp/script.js]: Not sure of functionality here, but even if N is strictly bigger than slides.length, will you still set it to zero? (or are you trying to do a circular wrap around?

Is it also possible to add unit tests here?

Comment (med0b1011) [reply portfolio/src/main/webapp/script.js]: To clarify, if n is slides.length + 1, do you need newindex to be 0 or 1 (and is that a valid case?)

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (PrabSG) [reply portfolio/src/main/webapp/script.js:88]: Added Jasmine using Karma to project, using Node. Added unit tests for the index calculation of this function (refactored first).

File: portfolio/src/main/webapp/script.js (hunk 3 above)
Comment (PrabSG) [reply portfolio/src/main/webapp/script.js]: Yes I noticed that too, after reconsidering what should be the correct functionality, I decided to make it more generic and wrap around by the same magnitude that it is greater/smaller by. So yes it is a circular wrap around, I have now refactored and slightly changed it to jump by two if say the index was two larger than the last item.

Comment (PrabSG) [reply portfolio/src/main/webapp/script.js]: Added unit tests for these cases also.

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -85,6 +79,17 @@ function showSlidePicture(n) {
   return newIndex;
 };
 
+/**
+ * Return correct index for a circular array.
+ */
+function circularIndex(n, items) {
```
Comment (med0b1011) [inline portfolio/src/main/webapp/script.js:115]: Would (n + items.length) % items.length work?

Also is it possible that this library already exists somewhere, and you can just use it? In this case you don't need to add tests. :)

Comment (med0b1011) [reply portfolio/src/main/webapp/script.js:115]: To be more generic, it can be more like

while (n < 0) n += items.length;

return n % items.length.

Depending how negative your offset can be.

Comment (PrabSG) [reply portfolio/src/main/webapp/script.js:115]: > To be more generic, it can be more like
> 
> while (n < 0) n += items.length;
> 
> return n % items.length.
> 
> Depending how negative your offset can be.

That is probably true that there is likely not going to be more than a few iterations of that loop. If the size is very small it could be possible to cause it to loop many times, should I consider that case? If so then it may be faster to do two cases.

Comment (PrabSG) [reply portfolio/src/main/webapp/script.js:115]: Have implemented the loop change as it simplifies logic and improves overall readability. Also this edge case is not possible in the current usage of these functions.

File: portfolio/test/main/webapp/script.spec.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+describe('Slide Pictures', () => {
+  
+  let items = []
+  
+  beforeEach(() => {
+    items = [0, 1, 2, 3];
```
Comment (DarthGandalf) [inline portfolio/test/main/webapp/script.spec.js]: items is not being changed, why do you need to reinitialize it every time?

Comment (PrabSG) [reply portfolio/test/main/webapp/script.spec.js]: Forgot to change this back. Done

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -31,16 +31,82 @@ function addRandomGreeting() {
...
+}
+
+/**
+ * Initialise map for currently displayed image.
+ */
+function initMap() {
+
+}
+
+let picCaptions = [
```
Comment (DarthGandalf) [inline portfolio/src/main/webapp/script.js]: Images are disconnected from captions. I think it would be better to either:
* have a single array with elements with 2 fields: src for the img and the caption
* put caption directly to html, inside slide-pic

Otherwise it's easy to accidentally reorder them, or add a new picture without adding its caption

Comment (PrabSG) [reply portfolio/src/main/webapp/script.js]: Good point, I have grouped the data to now be in a single array of objects, which can be future expanded depending on any other data that needs to be stored, or other images need to be added. To accommodate for this, HTML for the images is now generated using createElement on load of the page.

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -31,16 +31,131 @@ function addRandomGreeting() {
...
+function setGalleryImages() {
+  const slideContainer = document.getElementsByClassName('slides-container')[0];
+  const nextBtn = document.getElementsByClassName('next-slide')[0];
+  
+  // If we are on Home page gallery will not be defined.
+  if (typeof slideContainer === 'undefined') {
+    return;
+  }
+
+  for (pic of picData) {
```
Comment (DarthGandalf) [inline portfolio/src/main/webapp/script.js:60]: how does this work without let?

**Review by DarthGandalf — APPROVED** [review]

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (PrabSG) [reply portfolio/src/main/webapp/script.js:60]: Const is referring to the object reference, as the reference isn't changing, but the attributes are being changed this is still okay. At least that is what my knowledge of JS is, I may be mistaken

Comment (DarthGandalf) [reply portfolio/src/main/webapp/script.js:60]: What I meant is, I expected this line to be either `let pic of..` or `const pic of..`, but I see neither here. Your understanding of consts in JS looks correct
